### PR TITLE
Use WeakMap as descendant cache

### PR DESCRIPTION
**Pull request by fb55**
Otherwise this leaks memory when precompiling functions.

cc @dbuezas — would be great if you could have a look.

### CODE REVIEW

File: lib/general.js [hunk 1]
```diff
@@ -1,6 +1,20 @@
-var Set = require("es6-set");
 var attributeFactory = require("./attributes.js");
 
+// Highly specialized WeakMap shim
+var WMap = typeof WeakMap === "undefined" ? (function(){
```
Comment (realityking) [inline lib/general.js]: While browser support is worse this seems to be a better fit for `WeakSet`.

**Comment by dbuezas** [comment]
Did you observe a mem leak? The Set should be garbage collected as soon as the compiled selector finishes it's life cycle (i.e  the css query returns)

**Comment by dbuezas** [comment]
WeakSet/WeakMap were a bit slower last time I checked

**Comment by dbuezas** [comment]
WeakSet/WeakMap were a bit slower last time I checked

**Comment by dbuezas** [comment]
Furthermore, WeakSets/WeakMaps only hold weak references to their elements/keys. In this case they are the Dom elements and they will stay around longer than the compiled selector function anyway.

According to this Sets are faster than WeakSets/arrays (https://jsperf.com/es6-array-vs-set/6) and my own "real life" tests with this lib were consistent with that. Also the intended use is better reflected by the Set data structure.

All in all, I am of the opinion that this doesn't need to be changed. Thanks for keeping me in the loop!

**Comment by fb55** [comment]
> Did you observe a mem leak?

css-select can be used to pre-compile selectors, which speeds up selects. Eg. [node-inline](http://github.com/fb55/node-inline/) heavily depends on this. The current solution leaks memory, which is why I'd like to replace it with something that doesn't.

> Maybe use WeakSet

👍

**Comment by dbuezas** [comment]
Oh, wasn't aware of that use case, it makes sense.

Maybe an lru cache is a good fallback (instead of an array), if not a bit
over engineered.

**Comment by fb55** [comment]
Decided to only use the cache if the environment supports weaksets out of the box. This means older environments will take a performance hit, which isn't too bad.

**Pull request merged**